docs/design.md

```diff
@@ -0,0 +1,156 @@
+the reasons for the difference. This is in an effort to provide clarity.
+
+When features from Hypper can be upstreamed into Helm they will be. In that way,
+Hypper can act as a testing ground for Helm. We do not expect that every feature
+will fit in upstream Helm.
+
+A second element in the difference is that Helm, as a package manager, does not
+handle environment deployment operations. Those are left to higher level tools
+like Helmfile (a push model like Ansible) and the Flux Helm Operator (a pull
+model like Chef). Helm expects higher level tools will use Helm and charts for
+operating in different environments.
+
+Hypper will move will make some assumptions that move it more towards the
+environment management. This is in a different way from Helmfile or the Flux
+Helm Operator but further than Helm will go.
+
+## Directory Structure
+
+The directory structure is inspired by Helm with the following layout.
+
+```
+.
+├── cmd
+│   └── hypper
+├── docs
+└── pkg
+    └── action 
+    └── ... 
+```
+
+- The _cmd/hypper_ directory contains the client application. This CLI for mac,
+  linux, and windows is both a functional application and reference application
+  of the library.
+- _docs_ contains documentation on the design and usage of hypper.
+- The _pkg_ directory contains a library meant to be imported by _cmd/hypper_ and
+  other applications. The functionality needed is implemented here.
+- _pkg/actions_ contains the main functionality used by _cmd/hypper_.
+
+This is the same layout as Helm itself.
+
+## Design Intent
+
+The following outline the initial major intentions of hypper.
+
+### Modern Terminal UI
+
+Colors and icons are often present in modern terminals. Sometimes referred to
+as eye candy, these features can improve user experience by drawing attention to
+important things.
+
+### Optional Dependencies
+
+In Helm there are two ways to handle dependencies.
+
+1. Explicitly declare other charts as dependencies. This will cause those charts
+   to be installed. When the templates are rendered they happen in the same
+   sandboxed environment.
+2. Install charts separately and use values to connect them to each other. For
+   example, install PostreSQL with one chart and then pass connection information
+   into another chart using a Secret or values that generate a Secret.
+
+Hypper intends to add a 3rd form of dependency via an optional dependency. For
+example, you are developing something that requires a prometheus CRD. You can
+depend on a chart that provides that CRD but does not provide prometheus. The
+CRD chart or your chart can declare an optional dependency on a chart that
+installs Prometheus. When using hypper to install your chart you will be asked
+about installing the optional dependencies. In development you can choose not
+to install the dependencies while in production you can choose to install them.
+
+The optional dependency handling will extend to all commands.
```

Comment:
How are we going to provide directives on which optional dependencies to be installed?
Interactive through the cli? A deployment manifest?

Reply:
> How are we going to provide directives on which optional dependencies to be installed? Interactive through the cli? A deployment manifest?

Annotations in Helm charts. Rancher already uses them as do others. This way it can be right there in the package.